### PR TITLE
URA-872 - add config parsing and monitoring

### PR DESCRIPTION
**Pull request by jethror1**
- add parsing config file and beginning of monitored uploading
- new functions:
  - `utils.sizeof_fmt` - format given bytes to human readable string
  - `utils.read_config` - read config file into dict
  - `s3_upload.verify_config` - verify config params are valid
  - `utils.get_runs_to_upload` - return all run dirs that are uploadable
  - `s3_upload.monitor_directories_for_upload` - main function for monitoring
  - all logging related functions in `log.py`
- changed `upload.check_bucket_exists` to take multiple buckets
- move single run upload log into `s3_upload.upload_single_run`
- fix broken tests and add new ones for new functionality

- - -
This change is [Reviewable](https://reviewable.io/reviews/eastgenomics/s3_upload/5)

### CODE REVIEW

File: s3_upload/s3_upload.py [hunk 1]
```diff
@@ -81,29 +90,176 @@ def parse_args() -> argparse.Namespace:
...
+        errors.append("max_threads must be an integer")
+
+    if not config.get("log_dir"):
+        errors.append("required parameter log_dir not defined")
+
+    if not config.get("monitor"):
+        errors.append("required parameter monitor not defined")
+
+    for idx, monitor in enumerate(config.get("monitor")):
+        for key in ["monitored_directories", "bucket", "remote_path"]:
```
Comment (fletchercr) [inline s3_upload/s3_upload.py:133]: Does there need to be type checks for each key too? Or is that too specific?

**Review by jethror1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/s3_upload/5)* status: 0 of 9 files reviewed, 1 unresolved discussion (waiting on @fletchercr)

___
*[`s3_upload/s3_upload.py` line 133 at r2](https://reviewable.io/reviews/eastgenomics/s3_upload/5#gh-1791483669:-O8fOAIR05VmlD-x8rLI:b1f7xwq) ([raw file](https://github.com/eastgenomics/s3_upload/blob/566b1bca37295ab4b848e329aec218f315bcd6f2/s3_upload/s3_upload.py#L133)):*
<details><summary><i>Previously, fletchercr wrote…</i></summary><blockquote>

Does there need to be type checks for each key too? Or is that too specific?
</blockquote></details>

good point, switched this to a dict with expected types

File: s3_upload/utils/upload.py [hunk 2]
```diff
@@ -29,35 +35,51 @@ def check_aws_access():
     botocore.exceptions.ClientError
         Raised when unable to connect to AWS
     """
+    log.info("Checking access to AWS")
     try:
         return list(boto3.Session().resource("s3").buckets.all())
     except s3_exceptions.ClientError as err:
```
Comment (fletchercr) [inline s3_upload/utils/upload.py:41]: I'm new to logging, but do exceptions need to be logged?

**Review by jethror1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/s3_upload/5)* status: 0 of 9 files reviewed, 2 unresolved discussions (waiting on @fletchercr)

___
*[`s3_upload/utils/upload.py` line 41 at r2](https://reviewable.io/reviews/eastgenomics/s3_upload/5#gh-1791549246:-O8fVc2q7k3fBc9GEz_Q:b7zk3hg) ([raw file](https://github.com/eastgenomics/s3_upload/blob/566b1bca37295ab4b848e329aec218f315bcd6f2/s3_upload/utils/upload.py#L41)):*
<details><summary><i>Previously, fletchercr wrote…</i></summary><blockquote>

I'm new to logging, but do exceptions need to be logged?
</blockquote></details>

I probably should add it to the logs yes, but it will also raise an error and stop here anyway.

The logging basically allows for configurable printing and also redirects to a log file at the same time, if you use a mix of `log.debug`, `log.info`, `log.error` etc then you can configure how much logging shows up by setting the log level (which I will expose later via an `--log_level` argument). 

I will come back later and add in everything to log and to what level later I think, it can be useful to throw debug logs to most functions and add in variables, then for debugging you can turn on the log.debug level and get a lot of verbose output.

Log levels are explained here: https://docs.python.org/3/library/logging.html#logging-levels

**Review by jethror1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/s3_upload/5)* status: 0 of 9 files reviewed, 2 unresolved discussions (waiting on @fletchercr)

___
*[`s3_upload/utils/upload.py` line 41 at r2](https://reviewable.io/reviews/eastgenomics/s3_upload/5#gh-1791549246:-O8fWs2N17IkxCwcJOW_:bswz99p) ([raw file](https://github.com/eastgenomics/s3_upload/blob/566b1bca37295ab4b848e329aec218f315bcd6f2/s3_upload/utils/upload.py#L41)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

I probably should add it to the logs yes, but it will also raise an error and stop here anyway.

The logging basically allows for configurable printing and also redirects to a log file at the same time, if you use a mix of `log.debug`, `log.info`, `log.error` etc then you can configure how much logging shows up by setting the log level (which I will expose later via an `--log_level` argument). 

I will come back later and add in everything to log and to what level later I think, it can be useful to throw debug logs to most functions and add in variables, then for debugging you can turn on the log.debug level and get a lot of verbose output.

Log levels are explained here: https://docs.python.org/3/library/logging.html#logging-levels
</blockquote></details>

have added some more logging to the function too

**Review by jethror1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/s3_upload/5)* status: 0 of 9 files reviewed, 3 unresolved discussions (waiting on @fletchercr)

___
*[`s3_upload/s3_upload.py` line 150 at r4](https://reviewable.io/reviews/eastgenomics/s3_upload/5#gh-1791572231:-O8faLkaAVigRfZFsIRZ:b-aaq0du) ([raw file](https://github.com/eastgenomics/s3_upload/blob/c96b9407370851c9d4832999ddfa905c0d6db90b/s3_upload/s3_upload.py#L150)):*
<details><summary><i>Previously, fletchercr wrote…</i></summary><blockquote>

would this not then be redundant because of the isinstance changes above?
</blockquote></details>

yep it is, removed

**Review by fletchercr — APPROVED** [review]

**Pull request merged**